drivers/bt_ll_nrfxlib/clock_control/clock_control_ll_nrfxlib.c

```diff
@@ -0,0 +1,210 @@
+
+	if (rate == NULL) {
+		return -EINVAL;
+	}
+
+	*rate = 32768;
+	return 0;
+}
+
+#if defined(CONFIG_USB) && defined(CONFIG_SOC_NRF52840)
+static inline void power_event_cb(nrf_power_event_t event)
+{
+	extern void usb_dc_nrfx_power_event_callback(nrf_power_event_t event);
+
+	usb_dc_nrfx_power_event_callback(event);
+}
+#endif
+
+void nrf_power_clock_isr(void)
+{
+#if defined(CONFIG_USB) && defined(CONFIG_SOC_NRF52840)
+	bool usb_detected, usb_pwr_rdy, usb_removed;
+
+	usb_detected = nrf_power_event_check(NRF_POWER_EVENT_USBDETECTED);
+	usb_pwr_rdy = nrf_power_event_check(NRF_POWER_EVENT_USBPWRRDY);
+	usb_removed = nrf_power_event_check(NRF_POWER_EVENT_USBREMOVED);
+
+	if (usb_detected) {
+		nrf_power_event_clear(NRF_POWER_EVENT_USBDETECTED);
+		power_event_cb(NRF_POWER_EVENT_USBDETECTED);
+	}
+
+	if (usb_pwr_rdy) {
+		nrf_power_event_clear(NRF_POWER_EVENT_USBPWRRDY);
+		power_event_cb(NRF_POWER_EVENT_USBPWRRDY);
+	}
+
+	if (usb_removed) {
+		nrf_power_event_clear(NRF_POWER_EVENT_USBREMOVED);
+		power_event_cb(NRF_POWER_EVENT_USBREMOVED);
+	}
+#endif
+	/* FIXME/NOTE: This handler is called _after_ the USB registers are
+	 * checked. The reason is that the BLE controller also checks these
+	 * registsers _and clears them_, but there is currently no event
+	 * propagated from the controller.
+	 */
+	ble_controller_POWER_CLOCK_IRQHandler();
+}
+
+static int clock_control_init(struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	/* No-op. Initialized in subsys/bluetooth/controller/hci_driver.c by
+	 * ble_init() at PRE_KERNEL_1. ble_init() will call
+	 * ble_controller_init() that in turn will start the LFCLK.
+	 */
+	IRQ_CONNECT(DT_NORDIC_NRF_CLOCK_0_IRQ_0,
+		    DT_NORDIC_NRF_CLOCK_0_IRQ_0_PRIORITY,
+		    nrf_power_clock_isr, 0, 0);
+	return 0;
+}
+
+static const struct clock_control_driver_api hf_clock_control_api = {
+	.on = hf_clock_start,
+	.off = hf_clock_stop,
+	.get_rate = hf_clock_get_rate,
+};
+
+DEVICE_AND_API_INIT(hf_clock,
+		    DT_NORDIC_NRF_CLOCK_0_LABEL "_16M",
+		    clock_control_init, NULL, NULL, PRE_KERNEL_1,
+		    CONFIG_KERNEL_INIT_PRIORITY_DEVICE, &hf_clock_control_api);
+
+/* LFCLK doesn't have stop function to replicate the nRF5 Power Clock driver
+ * behavior.
+ */
+static const struct clock_control_driver_api lf_clock_control_api = {
+	.on = lf_clock_start,
+	.off = NULL,
+	.get_rate = lf_clock_get_rate,
+};
+
+DEVICE_AND_API_INIT(lf_clock,
+		    DT_NORDIC_NRF_CLOCK_0_LABEL "_32K",
+		    clock_control_init, NULL, NULL, PRE_KERNEL_1,
+		    CONFIG_KERNEL_INIT_PRIORITY_DEVICE, &lf_clock_control_api);
+
+#if defined(CONFIG_USB) && defined(CONFIG_SOC_NRF52840)
+#include <nrf_power.h>
+
+void nrf5_power_usb_power_int_enable(bool enable)
+{
+	u32_t mask;
+
+
+	mask = NRF_POWER_INT_USBDETECTED_MASK |
+	       NRF_POWER_INT_USBREMOVED_MASK |
+	       NRF_POWER_INT_USBPWRRDY_MASK;
+
+	if (enable) {
+		nrf_power_int_enable(mask);
+		irq_enable(DT_NORDIC_NRF_CLOCK_0_IRQ_0);
+	} else {
+		nrf_power_int_disable(mask);
```

Comment:
This function is taken directly from the in-tree driver, so I guess this is expected. It enables or disables the power int for the given mask.

Reply:
I see. I checked the code and you are right. Probably somebody found it easier to simply leave the irq enabled instead of maintaining the state (checking if it should be kept enabled even if usb disables it). I doubt this can be really a problem even if both usb and ble will be disabled at some point. Let's leave it as is. If somebody other then me thinks this is a problem we can create Jira ticket and fix it later.